In [1]:
import pandas as pd
import numpy as np

In [2]:
credits = pd.read_csv('credits.csv')

In [3]:
movies = pd.read_csv('movies_metadata.csv')
ratings = pd.read_csv('ratings_small.csv')
users = pd.read_csv('users.csv')

C:\Users\HARSH\AppData\Local\Temp\ipykernel_8768\1543059564.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  movies = pd.read_csv('movies_metadata.csv')


In [56]:
users.head(1)

userid   fname      lname   username gender            email   password  \
0       1  Rashad  Castaneda  Rashad@61   Male  Salvador@et.com   6fVt6H@    

  confirmpassword  
0        6fVt6H@

In [4]:
credits['id']=credits['id'].astype(str)

In [5]:
movies.isnull().sum()

adult                        0
belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
tagline                  25054
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

In [6]:
movies=movies.merge(credits,on='id')

In [7]:
movies.isnull().sum()

adult                        0
belongs_to_collection    41038
budget                       0
genres                       0
homepage                 37746
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   3
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      3
runtime                    260
spoken_languages             3
status                      84
tagline                  25099
title                        3
video                        3
vote_average                 3
vote_count                   3
cast                         0
crew                         0
dtype: int64

In [8]:
movies = movies[['belongs_to_collection','genres','id','original_title','overview','cast','crew']]

In [9]:
movies.isnull().sum()

belongs_to_collection    41038
genres                       0
id                           0
original_title               0
overview                   954
cast                         0
crew                         0
dtype: int64

In [10]:
movies.duplicated().sum()

99

In [11]:
movies.dropna(inplace=True)

In [12]:
movies.duplicated().sum()

11

In [13]:
movies.drop_duplicates()

belongs_to_collection  \
0      [{'id': 10194, 'name': 'Toy Story Collection',...   
2      [{'id': 119050, 'name': 'Grumpy Old Men Collec...   
4      [{'id': 96871, 'name': 'Father of the Bride Co...   
9      [{'id': 645, 'name': 'James Bond Collection', ...   
12     [{'id': 117693, 'name': 'Balto Collection', 'p...   
...                                                  ...   
45427  [{'id': 37261, 'name': 'The Carry On Collectio...   
45430  [{'id': 37261, 'name': 'The Carry On Collectio...   
45441  [{'id': 37261, 'name': 'The Carry On Collectio...   
45443  [{'id': 477208, 'name': 'DC Super Hero Girls C...   
45454  [{'id': 200641, 'name': 'Red Lotus Collection'...   

                                                  genres      id  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...     862   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   15602   
4                         [{'id': 35, 'name': 'Comedy'}]   11862   
9      [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...     710   
12     [{'id': 10751, 'name': 'Family'}, {'id': 16, '...   21032   
...                                                  ...     ...   
45427                     [{'id': 35, 'name': 'Comedy'}]   24568   
45430                     [{'id': 35, 'name': 'Comedy'}]   19307   
45441                     [{'id': 35, 'name': 'Comedy'}]   21251   
45443                  [{'id': 16, 'name': 'Animation'}]  460135   
45454  [{'id': 28, 'name': 'Action'}, {'id': 10769, '...  182030   

                              original_title  \
0                                  Toy Story   
2                           Grumpier Old Men   
4                Father of the Bride Part II   
9                                  GoldenEye   
12                                     Balto   
...                                      ...   
45427                      Follow That Camel   
45430                       Carry On Camping   
45441                       Carry On England   
45443  LEGO DC Super Hero Girls: Brain Drain   
45454                             火燒紅蓮寺之鴛鴦劍俠   

                                                overview  \
0      Led by Woody, Andy's toys live happily in his ...   
2      A family wedding reignites the ancient feud be...   
4      Just when George Banks has recovered from his ...   
9      James Bond must unmask the mysterious head of ...   
12     An outcast half-wolf risks his life to prevent...   
...                                                  ...   
45427  Bertram Oliphant West (also known as Bo West) ...   
45430  Sid and Bernie keep having their amorous inten...   
45441  Captain S. Melly takes over as the new Command...   
45443  When Supergirl, Wonder Woman, Batgirl, Bumbleb...   
45454  Kwei Wu happens upon a kidnapping with his wif...   

                                                    cast  \
0      [{'cast_id': 14, 'character': 'Woody (voice)',...   
2      [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
4      [{'cast_id': 1, 'character': 'George Banks', '...   
9      [{'cast_id': 1, 'character': 'James Bond', 'cr...   
12     [{'cast_id': 1, 'character': 'Balto (voice)', ...   
...                                                  ...   
45427  [{'cast_id': 3, 'character': 'Sergeant Nocker'...   
45430  [{'cast_id': 3, 'character': 'Sid Boggle', 'cr...   
45441  [{'cast_id': 4, 'character': 'Captain S. Melly...   
45443  [{'cast_id': 0, 'character': 'Wonder Woman', '...   
45454  [{'cast_id': 4, 'character': 'Gui Wu', 'credit...   

                                                    crew  
0      [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  
2      [{'credit_id': '52fe466a9251416c75077a89', 'de...  
4      [{'credit_id': '52fe44959251416c75039ed7', 'de...  
9      [{'credit_id': '52fe426ec3a36847f801e14b', 'de...  
12     [{'credit_id': '593f24b9c3a3680369002371', 'de...  
...                                                  ...  
45427  [{'credit_id': '52fe449dc3a368484e02d0f7', 'de...  
45430 

In [14]:
movies.shape

(4441, 7)

In [15]:
import ast

In [16]:
def convert(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

In [17]:
movies['genres']=movies['genres'].apply(convert)

In [18]:
movies['cast']=movies['cast'].apply(convert)

In [19]:
def fetch_director(obj):
    L = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            L.append(i['name'])
            break
    return L

In [20]:
movies['crew']=movies['crew'].apply(fetch_director)

In [21]:
movies.head(1)

belongs_to_collection  \
0  [{'id': 10194, 'name': 'Toy Story Collection',...   

                        genres   id original_title  \
0  [Animation, Comedy, Family]  862      Toy Story   

                                            overview  \
0  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]

In [22]:
import ast
def convert1(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['poster_path'])
    return L

In [23]:
movies['belongs_to_collection']=movies['belongs_to_collection'].apply(convert1)

In [24]:
movies.head(1)

belongs_to_collection                       genres   id  \
0  [/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg]  [Animation, Comedy, Family]  862   

  original_title                                           overview  \
0      Toy Story  Led by Woody, Andy's toys live happily in his ...   

                                                cast             crew  
0  [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...  [John Lasseter]

In [25]:
ratings = ratings[['userId','movieId','rating']]

In [26]:
ratings['id']=ratings['movieId']

In [27]:
ratings['id']=ratings['id'].astype(str)

In [28]:
ratings.head(1)

userId  movieId  rating  id
0       1       31     2.5  31

In [29]:
ratings_with_id=ratings.merge(movies,on='id')

In [30]:
num_rating_df=ratings_with_id.groupby('id').count()['rating'].reset_index()
num_rating_df.rename(columns={'rating':'num_ratings'},inplace=True)
num_rating_df

id  num_ratings
0    100450            1
1       105           48
2      1051            7
3       108            1
4      1091           27
..      ...          ...
532      96            1
533     966            1
534      97            8
535     984            2
536     990            1

[537 rows x 2 columns]

In [31]:
avg_rating_df=ratings_with_id.groupby('id')['rating'].mean().rename('avg_rating')
avg_rating_df

id
100450    3.500000
105       3.114583
1051      3.000000
108       4.000000
1091      2.629630
            ...   
96        1.000000
966       4.000000
97        3.875000
984       4.500000
990       2.500000
Name: avg_rating, Length: 537, dtype: float64

In [32]:
popular_df = num_rating_df.merge(avg_rating_df,on='id')
popular_df

id  num_ratings  avg_rating
0    100450            1    3.500000
1       105           48    3.114583
2      1051            7    3.000000
3       108            1    4.000000
4      1091           27    2.629630
..      ...          ...         ...
532      96            1    1.000000
533     966            1    4.000000
534      97            8    3.875000
535     984            2    4.500000
536     990            1    2.500000

[537 rows x 3 columns]

In [33]:
popular_df = popular_df[popular_df['num_ratings']>=50].sort_values('avg_rating',ascending=False).head(50)
popular_df

id  num_ratings  avg_rating
354    527          244    4.303279
130   2019           54    4.277778
399    608          224    4.256696
226    296          324    4.256173
516    923           85    4.200000
528    953           72    4.152778
15    1250           65    4.138462
515    920           68    4.044118
313  44191           73    4.013699
166   2395           68    3.963235
324    457          213    3.953052
62      16           88    3.948864
6      110          228    3.945175
178   2502          106    3.915094
47     150          200    3.902500
80    1682          137    3.886861
17    1266           57    3.885965
408    628           50    3.880000
506   8961          126    3.861111
63     161           80    3.831250
220   2791          106    3.820755
24    1285           61    3.803279
115   1954           69    3.789855
13    1246           95    3.784211
272    364          200    3.777500
32    1374           72    3.750000
343   5060           59    3.745763
175     25          101    3.742574
139   2109           51    3.696078
403     62           87    3.689655
5       11           82    3.689024
135   2080           51    3.686275
128   2005           51    3.676471
395    588          215    3.674419
369    551           90    3.650000
414    661           53    3.632075
36     141           93    3.580645
68    1639           65    3.569231
276    377          180    3.566667
312    440           90    3.561111
137     21           95    3.536842
67     163           63    3.515873
250   3176           66    3.500000
370    553           57    3.482456
338    494           61    3.475410
56    1573           83    3.451807
72     165          142    3.443662
361    539          125    3.428000
431    708           61    3.401639
179    253          125    3.396000

In [34]:
popular_df.shape

(50, 3)

In [35]:
movies['belongs_to_collection']=movies['belongs_to_collection'].str[0]

In [36]:
movies.head()

belongs_to_collection                          genres     id  \
0   /7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg     [Animation, Comedy, Family]    862   
2   /nLvUdqgPgm3F85NMCii9gVFUcet.jpg               [Romance, Comedy]  15602   
4   /nts4iOmNnq7GNicycMJ9pSAn204.jpg                        [Comedy]  11862   
9    /HORpg5CSkmeQlAolx3bKMrKgfi.jpg   [Adventure, Action, Thriller]    710   
12  /w0ZgH6Lgxt2bQYnf1ss74UvYftm.jpg  [Family, Animation, Adventure]  21032   

                 original_title  \
0                     Toy Story   
2              Grumpier Old Men   
4   Father of the Bride Part II   
9                     GoldenEye   
12                        Balto   

                                             overview  \
0   Led by Woody, Andy's toys live happily in his ...   
2   A family wedding reignites the ancient feud be...   
4   Just when George Banks has recovered from his ...   
9   James Bond must unmask the mysterious head of ...   
12  An outcast half-wolf risks his life to prevent...   

                                                 cast               crew  
0   [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...    [John Lasseter]  
2   [Walter Matthau, Jack Lemmon, Ann-Margret, Sop...    [Howard Deutch]  
4   [Steve Martin, Diane Keaton, Martin Short, Kim...    [Charles Shyer]  
9   [Pierce Brosnan, Sean Bean, Izabella Scorupco,...  [Martin Campbell]  
12  [Kevin Bacon, Bob Hoskins, Bridget Fonda, Jim ...      [Simon Wells]

In [37]:
ratings_with_id.head(1)

userId  movieId  rating    id               belongs_to_collection   genres  \
0       1     1371     2.5  1371  [/mCY5dMkSSFQufGCViI6jNUU6pXq.jpg]  [Drama]   

  original_title                                           overview  \
0      Rocky III  Now the world champion, Rocky Balboa is living...   

                                                cast                  crew  
0  [Sylvester Stallone, Talia Shire, Burt Young, ...  [Sylvester Stallone]

users who gave minimum 30 ratings to movies...... 

In [38]:
x=ratings_with_id.groupby('userId').count()['rating'] > 20
litrate_user = x[x].index

out of all user's who had rated the movies...this will filtered the user's who are litrate....

In [39]:
filtered_rating = ratings_with_id[ratings_with_id['userId'].isin(litrate_user)]

movies that have minimum 30 ratings....

In [40]:
y = filtered_rating.groupby('original_title').count()['rating']>=20
famous_books = y[y].index

out of all movies.. this will filtered the movies that have 30 or more ratings...

In [41]:
final_ratings = filtered_rating[filtered_rating['original_title'].isin(famous_books)]

In [42]:
final_ratings.shape

(4495, 10)

creating a pivot tables....

In [43]:
pt = final_ratings.pivot_table(index='original_title',columns='userId',values='rating') 
pt

userId                    4    15   17   19   21   23   30   34   36   48   \
original_title                                                               
2001: A Space Odyssey     NaN  2.0  NaN  NaN  NaN  4.0  NaN  NaN  NaN  NaN   
28 Days Later             NaN  2.5  4.0  NaN  NaN  NaN  4.0  NaN  NaN  NaN   
28 Weeks Later            NaN  1.5  NaN  NaN  NaN  1.0  3.0  NaN  NaN  NaN   
300                       NaN  NaN  NaN  3.0  NaN  NaN  4.0  NaN  NaN  NaN   
48 Hrs.                   NaN  3.0  NaN  3.0  NaN  3.5  5.0  4.0  NaN  NaN   
...                       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
Trois couleurs : Rouge    NaN  3.0  NaN  3.0  NaN  3.5  5.0  4.0  NaN  4.0   
X-Men Origins: Wolverine  5.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
バトル・ロワイアル                 NaN  2.0  NaN  NaN  NaN  5.0  4.0  NaN  NaN  NaN   
リング                       NaN  4.0  NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN   
怪獣総進撃                     NaN  3.0  NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN   

userId                    ...  641  647  648  649  652  654  659  664  665  \
original_title            ...                                                
2001: A Space Odyssey     ...  4.0  NaN  NaN  NaN  NaN  4.5  3.0  NaN  3.0   
28 Days Later             ...  NaN  3.0  NaN  NaN  NaN  NaN  NaN  3.0  NaN   
28 Weeks Later            ...  NaN  NaN  NaN  NaN  4.0  3.0  NaN  NaN  NaN   
300                       ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
48 Hrs.                   ...  4.0  NaN  NaN  4.0  NaN  5.0  5.0  NaN  NaN   
...                       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
Trois couleurs : Rouge    ...  NaN  NaN  NaN  NaN  NaN  4.5  3.0  4.0  NaN   
X-Men Origins: Wolverine  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  4.0   
バトル・ロワイアル                 ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
リング                       ...  NaN  NaN  3.5  NaN  3.5  3.5  NaN  NaN  3.0   
怪獣総進撃                     ...  NaN  NaN  NaN  NaN  NaN  4.5  NaN  NaN  NaN   

userId                    667  
original_title                 
2001: A Space Odyssey     NaN  
28 Days Later             NaN  
28 Weeks Later            NaN  
300                       NaN  
48 Hrs.                   4.0  
...                       ...  
Trois couleurs : Rouge    4.0  
X-Men Origins: Wolverine  NaN  
バトル・ロワイアル                 NaN  
リング                       NaN  
怪獣総進撃                     NaN  

[108 rows x 152 columns]

In [44]:
pt.fillna(0,inplace=True)
pt

userId                    4    15   17   19   21   23   30   34   36   48   \
original_title                                                               
2001: A Space Odyssey     0.0  2.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0  0.0   
28 Days Later             0.0  2.5  4.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0   
28 Weeks Later            0.0  1.5  0.0  0.0  0.0  1.0  3.0  0.0  0.0  0.0   
300                       0.0  0.0  0.0  3.0  0.0  0.0  4.0  0.0  0.0  0.0   
48 Hrs.                   0.0  3.0  0.0  3.0  0.0  3.5  5.0  4.0  0.0  0.0   
...                       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
Trois couleurs : Rouge    0.0  3.0  0.0  3.0  0.0  3.5  5.0  4.0  0.0  4.0   
X-Men Origins: Wolverine  5.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
バトル・ロワイアル                 0.0  2.0  0.0  0.0  0.0  5.0  4.0  0.0  0.0  0.0   
リング                       0.0  4.0  0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0   
怪獣総進撃                     0.0  3.0  0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0   

userId                    ...  641  647  648  649  652  654  659  664  665  \
original_title            ...                                                
2001: A Space Odyssey     ...  4.0  0.0  0.0  0.0  0.0  4.5  3.0  0.0  3.0   
28 Days Later             ...  0.0  3.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0   
28 Weeks Later            ...  0.0  0.0  0.0  0.0  4.0  3.0  0.0  0.0  0.0   
300                       ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
48 Hrs.                   ...  4.0  0.0  0.0  4.0  0.0  5.0  5.0  0.0  0.0   
...                       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
Trois couleurs : Rouge    ...  0.0  0.0  0.0  0.0  0.0  4.5  3.0  4.0  0.0   
X-Men Origins: Wolverine  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  4.0   
バトル・ロワイアル                 ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
リング                       ...  0.0  0.0  3.5  0.0  3.5  3.5  0.0  0.0  3.0   
怪獣総進撃                     ...  0.0  0.0  0.0  0.0  0.0  4.5  0.0  0.0  0.0   

userId                    667  
original_title                 
2001: A Space Odyssey     0.0  
28 Days Later             0.0  
28 Weeks Later            0.0  
300                       0.0  
48 Hrs.                   4.0  
...                       ...  
Trois couleurs : Rouge    4.0  
X-Men Origins: Wolverine  0.0  
バトル・ロワイアル                 0.0  
リング                       0.0  
怪獣総進撃                     0.0  

[108 rows x 152 columns]

Identifying Uclidian distance of one movies with other 108 movies..... 

In [45]:
from sklearn.metrics.pairwise import cosine_similarity

c:\users\harsh\python\python39\lib\site-packages\scipy\__init__.py:138: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.2)
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion} is required for this version of "


In [46]:
similarity_scores = cosine_similarity(pt)

Function that identifies the 6 movies related to movie title which uclidian distance is nearer....

In [123]:
index2 = np.where(pt.index=="X-Men Origins: Wolverine")[0][0]
index2

similar_items1 = sorted(list(enumerate(similarity_scores[index2])),key=lambda x:x[1],reverse=True)[1:6]
similar_items1

[(78, 0.5150697815166194),
 (54, 0.4987399105460614),
 (25, 0.4861207603313559),
 (20, 0.485848071148946),
 (61, 0.47827086997043883)]

In [121]:
def recommend(movie_name):
    #index fetch
    index1 = np.where(pt.index==movie_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_scores[index1])),key=lambda x:x[1],reverse=True)[1:6]
    
    data = []
    for i in similar_items:
        item = []
        temp_df = movies[movies['original_title'] == pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('original_title')['original_title'].values))
        item.extend(list(temp_df.drop_duplicates('original_title')['crew'].values))
        item.extend(list(temp_df.drop_duplicates('original_title')['belongs_to_collection'].values))
        
        data.append(item)
        
    return data

In [68]:
recommend('2001: A Space Odyssey')

[['Psycho', ['Alfred Hitchcock'], '/29pDqAgmb4idzWzuZncoT7LZplA.jpg'],
 ['Donnie Darko', ['Richard Kelly'], '/4Oj6ctBucsLvkbKsQMDaDXXR5cu.jpg'],
 ['Star Wars', ['George Lucas'], '/ghd5zOQnDaDW1mxO7R5fXXpZMu.jpg'],
 ["Rosemary's Baby", ['Roman Polanski'], '/yP9DnKsSGcEuHJktQYotND6dsgJ.jpg'],
 ['Trois couleurs : Rouge',
  ['Krzysztof Kieślowski'],
  '/rVdd23QuT5rHX7lZvuAkRRUkeCZ.jpg']]

In [55]:
import pickle
pickle.dump(pt,open('pt.pkl','wb'))
pickle.dump(movies,open('movies.pkl','wb'))
pickle.dump(similarity_scores,open('similarity_scores.pkl','wb'))